mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -1,115 +1,139 @@
+            Ok(Authorization::Forbidden(format!(
+                "{} is forbidden topic filter",
+                topic
+            )))
         } else {
-            // allow any client to publish to any non-iothub topics
-            Authorization::Allowed
+            // delegate to inner authorizer for to any non-iothub topics.
+            self.inner.authorize(activity)
         }
     }
 
     fn authorize_subscribe(
         &self,
-        client_id: &ClientId,
-        client_info: &ClientInfo,
+        activity: &Activity,
         subscribe: &Subscribe,
-    ) -> Authorization {
+    ) -> Result<Authorization, E> {
         let topic = subscribe.topic_filter();
 
         if is_iothub_topic(topic) {
             // run authorization rules for subscription to IoTHub topic
-            self.authorize_iothub_topic(client_id, client_info, topic)
+            self.authorize_iothub_topic(activity, topic)
         } else if is_forbidden_topic_filter(topic) {
             // forbid any clients to access restricted topics
-            Authorization::Forbidden(format!("{} is forbidden topic filter", topic))
+            Ok(Authorization::Forbidden(format!(
+                "{} is forbidden topic filter",
+                topic
+            )))
         } else {
-            // allow any client to subscribe to any non-iothub topics
-            Authorization::Allowed
+            // delegate to inner authorizer for to any non-iothub topics.
+            self.inner.authorize(activity)
         }
     }
 
-    fn authorize_iothub_topic(
-        &self,
-        client_id: &ClientId,
-        client_info: &ClientInfo,
-        topic: &str,
-    ) -> Authorization {
-        match client_info.auth_id() {
+    fn authorize_iothub_topic(&self, activity: &Activity, topic: &str) -> Result<Authorization, E> {
+        Ok(match activity.client_info().auth_id() {
             // forbid anonymous clients to subscribe to restricted topics
             AuthId::Anonymous => Authorization::Forbidden(
                 "Anonymous clients do not have access to IoTHub topics".to_string(),
             ),
             // allow authenticated clients with client_id == auth_id and accessing its own IoTHub topic
-            AuthId::Identity(identity) if identity == client_id => {
-                if self.is_iothub_topic_allowed(client_id, topic)
-                    && self.check_authorized_cache(client_id, topic)
+            AuthId::Identity(identity) if identity == activity.client_id() => {
+                if self.is_iothub_topic_allowed(activity.client_id(), topic)
+                    && self.check_authorized_cache(activity.client_id(), topic)
                 {
                     Authorization::Allowed
                 } else {
-                    Authorization::Forbidden(
-                        "Client must connect to its own IoTHub topic".to_string(),
-                    )
+                    // check if iothub policy is overridden by a custom policy.
```

Comment:
Are custom policies allowed to override iothub policies? Or are they only supposed to override in the case that iothub policy is not authorized? 

For example, can custom policy deny something that iothub policy allows? If that is true, then this won't work for that case. The auth will return Allowed

Reply:
We decided on two key scenarios:
- Customers can override IoT Hub authorization to give devices (modules) additional access to IoT Hub primitives. 
- Customers cannot override IoT Hub authorization to remove devices (modules) access to their own IoT Hub primitives.